### DISCUSSION TITLE
[Layout] Allowing outer layout properties on elements

### DISCUSSION THREAD

**Discussion by yurm04** [Shopify/polaris-react#7630] [Ideas]
## TL;DR
Should we add a `display` prop to some of our components so that they can set their outer layout?

## Context
[HTML elements have two display types](https://developer.mozilla.org/en-US/docs/Web/CSS/display#specifications): _inner_ and _outer_. Inner display types set the layout of an element's children (ex. `grid` and `flex`). Our new foundational layout components seem to mostly deal with setting the component's children (ex. `Inline`, `Grid`, `Tiles`)

Outer display types set the layout of the element itself (ex. `inline` and `block`). Right now, we're not able to set the outer display type of our components. There are also _precomposed_ display types that set the inner and outer properties together `display: inline-flex`. You can also use the syntax `display: inline flex`, but this is not well supported in all browsers just yet.

So, the question is: do we think we'll need or want to be able to set an element's outer display type directly?

## Prototype
I prototyped allowing a `display` property on the `Box` components to set the `display` CSS property [on this branch](https://github.com/Shopify/polaris/blob/prototype-box-display-property/polaris-react/src/components/Box/Box.tsx#L138-L164). I copied a list of all display values, but if we want to allow this prop then I think we can narrow this down only to the outer display types. This will mean that users will still need to use the new Layout components in most cases.

Something more like this:
```tsx
export type Display =
  | 'block'
  | 'inline'
  | 'inline-block'
  | 'inline-flex'
  | 'inline-grid'
```

## Background
I stumbled on this question while trying to rebuild the `Tag` components using our new layout components. The `Tag` components can render a few different ways: as a [`<span>`](https://github.com/Shopify/polaris/blob/main/polaris-react/src/components/Tag/Tag.tsx#L105-L108) with nested [`a` or `span`](https://github.com/Shopify/polaris/blob/main/polaris-react/src/components/Tag/Tag.tsx#L90-L101) children and an optional inline [`button`](https://github.com/Shopify/polaris/blob/main/polaris-react/src/components/Tag/Tag.tsx#L107) child. Or it can return a [`button`](https://github.com/Shopify/polaris/blob/main/polaris-react/src/components/Tag/Tag.tsx#L52-L63). 

The `Tag` is currently styled using [`display: inline-flex`](https://github.com/Shopify/polaris/blob/main/polaris-react/src/components/Tag/Tag.scss#L7), which sets both the outer and inner display type for the rendered element. Trying to recreate this solution using the `Inline` and `Box` components and `span` isn't working because we can't set the height of an inline element.

`span` rendered from `Box`
<img width="97" alt="Screen Shot 2022-11-03 at 9 53 15 AM" src="https://user-images.githubusercontent.com/4642404/199756533-408e8e2f-b499-4ded-beb5-8fa6961b4142.png">

`button` rendered inside `span`
<img width="107" alt="Screen Shot 2022-11-03 at 9 53 44 AM" src="https://user-images.githubusercontent.com/4642404/199756657-3a10a2fb-7663-481c-99eb-4d909a904cda.png">

This led me down the rabbit hole of of inner and outer display types described above.

**Comment by aveline** [comment]
Some initial thoughts, perhaps we are missing a layout primitive that would handle this?

By adding `display` to `Box`, do we turn it into a component also used for layout? Currently my understanding is that `Box` is meant to be a low level way to access design tokens and design opinions.

**Comment by kyledurand** [comment]
> Currently my understand is that Box is meant to be a low level way to access design tokens and design opinions.

I like this description of what Box is meant to be. I'm not sure we're missing a primitive but perhaps these could be props to on our layout components vs on Box itself. I could see `Inline` taking a prop that turns it into an inline-flex container, same with columns and `inline-grid`. 

I'd like to make sure we have a definitive use case for adding this kind of feature to these components before going ahead with this though
